### PR TITLE
Introduce and use a _StringGuts visitor

### PR DESCRIPTION
**Pull request by milseman**
Continuation of https://github.com/apple/swift/pull/14470. Starting to get a clean and consistent introspection pattern that allows us to make much of _StringGuts's bit pattern resilient.

### CODE REVIEW

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please test compiler performance

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
This pattern still leaves much to be desired.

* The use of the visitor needs to make sure their closures end up being thin, otherwise there's a severe perf cost there. Since this is all internal, I'd like some way to assert on the thinness of the closure to detect perf regressions earlier.

* The closures not being declared thin means that we waste 3 registers, so x86_64 has to spill for some of its arguments (and arm64 too, but for an additional reason below). I'd like to unsafe cast them to thin to save 3 registers.

* Trying to unify ranged and non-ranged visitors does cause extra register use across the opaque visit call, which could mean more spillage and/or saving of registers. I think I should duplicate the private visit functions to work around this.

* This is still ARC-fragile, in that we don't have any guarantees that there won't be excessive ARC caused by this, just that the current code pattern happens to not ARC, likely do to a peculiarity of SILGen and not even the ARC optimizer.

**Comment by milseman** [comment]
The above restore Join's performance, but not yet the others.

**Comment by milseman** [comment]
More points:

* I could use @convention(thin) if I can force the generic types into a unified representation. This could be by passing them indirectly, monomorphize with gyb, etc.

File: stdlib/public/core/StringGutsVisitor.swift [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+  @inline(never)
+  func _visitOpaque<Result>(
+    range: (Range<Int>, performBoundsCheck: Bool)? = nil,
+    ascii: /*@convention(thin)*/ (_UnmanagedString<UInt8>) -> Result,
+    utf16: /*@convention(thin)*/ (_UnmanagedString<UInt16>) -> Result,
+    opaque: /*@convention(thin)*/ (_UnmanagedOpaqueString) -> Result
+  ) -> Result {
+    _sanityCheck(_guts._isOpaque)
+
+    // TODO: But can it provide a pointer+length representation?
```
Comment (lorentey) [inline stdlib/public/core/StringGutsVisitor.swift]: Ah, so e.g., we can have the SSO case opaque, but still call the ASCII/UTF-16 closures, right? 👌

File: stdlib/public/core/StringGuts.swift [hunk 2]
```diff
@@ -572,8 +572,7 @@ extension _StringGuts {
 }
 
 extension _StringGuts {
-  @inline(never)
-  @_versioned
```
Comment (lorentey) [inline stdlib/public/core/StringGuts.swift]: 🎉

**Comment by lorentey** [comment]
I have an abandoned branch somewhere that replaces the `_UnmanagedString` generic with two concrete types plus an internal protocol. (It seemed like a good idea to do that purely for code readability, even without gybbing.) Would doing that help here?

**Comment by milseman** [comment]
Do you mean that I could unify some of these visitors's `ascii` and `utf16` closures by using the existential? It might help, assuming that the existential always gets totally de-virtualized. However, the I'm a little worried about relying on it throughout the rest of the code.

**Comment by milseman** [comment]
So the main cause of `CSVParsingAltIndices` slowdown is `UnicodeScalarView.index(after:)`. I inspected the compiled code of that method before and after and it's nearly identical, even down to the code layout of the taken-path. 

The only difference is the larger stack frame and potentially higher register pressure (though the taken path didn't differ in spilling). I'll have to investigate more on the actual benchmark compiled code and/or tune the visitor anyways to reduce register utilization.

**Pull request closed without merging**